client/fs_endpoint.go

```diff
@@ -0,0 +1,921 @@
+
+	// Decode the arguments
+	var req cstructs.FsStreamRequest
+	decoder := codec.NewDecoder(conn, structs.MsgpackHandle)
+	encoder := codec.NewEncoder(conn, structs.MsgpackHandle)
+
+	if err := decoder.Decode(&req); err != nil {
+		f.handleStreamResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	// Check read permissions
+	if aclObj, err := f.c.ResolveToken(req.QueryOptions.AuthToken); err != nil {
+		f.handleStreamResultError(err, nil, encoder)
+		return
+	} else if aclObj != nil && !aclObj.AllowNsOp(req.Namespace, acl.NamespaceCapabilityReadFS) {
+		f.handleStreamResultError(structs.ErrPermissionDenied, nil, encoder)
+		return
+	}
+
+	// Validate the arguments
+	if req.AllocID == "" {
+		f.handleStreamResultError(allocIDNotPresentErr, helper.Int64ToPtr(400), encoder)
+		return
+	}
+	if req.Path == "" {
+		f.handleStreamResultError(pathNotPresentErr, helper.Int64ToPtr(400), encoder)
+		return
+	}
+	switch req.Origin {
+	case "start", "end":
+	case "":
+		req.Origin = "start"
+	default:
+		f.handleStreamResultError(invalidOrigin, helper.Int64ToPtr(400), encoder)
+		return
+	}
+
+	fs, err := f.c.GetAllocFS(req.AllocID)
+	if err != nil {
+		var code *int64
+		if strings.Contains(err.Error(), "unknown allocation") {
+			code = helper.Int64ToPtr(404)
+		} else {
+			code = helper.Int64ToPtr(500)
+		}
+
+		f.handleStreamResultError(err, code, encoder)
+		return
+	}
+
+	// Calculate the offset
+	fileInfo, err := fs.Stat(req.Path)
+	if err != nil {
+		f.handleStreamResultError(err, helper.Int64ToPtr(400), encoder)
+		return
+	}
+	if fileInfo.IsDir {
+		f.handleStreamResultError(
+			fmt.Errorf("file %q is a directory", req.Path),
+			helper.Int64ToPtr(400), encoder)
+		return
+	}
+
+	// If offsetting from the end subtract from the size
+	if req.Origin == "end" {
+		req.Offset = fileInfo.Size - req.Offset
+		if req.Offset < 0 {
+			req.Offset = 0
+		}
+	}
+
+	frames := make(chan *sframer.StreamFrame, streamFramesBuffer)
+	errCh := make(chan error)
+	var buf bytes.Buffer
+	frameCodec := codec.NewEncoder(&buf, structs.JsonHandle)
+
+	// Create the framer
+	framer := sframer.NewStreamFramer(frames, streamHeartbeatRate, streamBatchWindow, streamFrameSize)
+	framer.Run()
+	defer framer.Destroy()
+
+	// If we aren't following end as soon as we hit EOF
+	var eofCancelCh chan error
+	if !req.Follow {
+		eofCancelCh = make(chan error)
+		close(eofCancelCh)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Start streaming
+	go func() {
+		if err := f.streamFile(ctx, req.Offset, req.Path, req.Limit, fs, framer, eofCancelCh); err != nil {
+			select {
+			case errCh <- err:
+			case <-ctx.Done():
+			}
+		}
+
+		framer.Destroy()
+	}()
+
+	// Create a goroutine to detect the remote side closing
+	go func() {
+		for {
+			if _, err := conn.Read(nil); err != nil {
+				if err == io.EOF {
+					cancel()
+					return
+				}
+				select {
+				case errCh <- err:
+				case <-ctx.Done():
+					return
```

Comment:
Shouldn't it close the connection from this end for these two cases?

Reply:
At the top of the function there is a defer close. So this will cancel the stream framer which causes the below select loop to break out and you exit all together once you've flushed.